### ISSUE TITLE
Telemetry: Report if command succeded or failed

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#9346]
### Use case description

It's valuable to know whether given command succeeded or failed. Currently payload as sent does not share this data

### Proposed solution

To be able to report success or failure, we need to store the command information at the end of the processing command and not at the end as it happens now. Switching that order may slightly prolong the command with which telemetry is send, still currently we send it only with `deploy` command, which already as it involves a full CloudFormation deploy behind is usually time taking, so that shouldn't influence the Framework DX.

Add to payload `outcome` property, which should be set to either `success` or `failure` property.

When the failure happens, report additional data with `failureReason` property object, which should have following properties:
- `kind` - Kind of an error, either `user` (error on user side) or `programmer` (error on our or plugin side)
- `code` - `error.code` property (should be set for all _user_ errors, may be set for some _programmer_ errors)
- `location` - First few lines of a stack trace but excluding the error message (message may contain some user created data, we don't want to send it). I think we should send it only for _programmer_ errors, as for _user_ errors, `code` should give us all needed information

### How to implement it?

_Let's address each top level point with individual PR_

1. [x] Ensure that all `ServerlessError` constructs, both in this repository and in `@serverless/enterprise-plugin` provide a meaningful _error code_ as second argument, also verify that currently used codes are good.
1. [x] Move logic with which we [store command telemetry, and eventually send it](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/lib/classes/PluginManager.js#L589), so it happens after _command_ is processed, and outside of `Serverless` instance realm (we need to do that, to be able to track well in a future e.g. interactive setup which may trigger multiple commands internally(e.g. `create` then `deploy`)). 
   - Add `isTelemetryReportedExternally: true` to [`Serverless` constructor options](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/scripts/serverless.js#L360-L367), and assign it to `serverless.isTelemetryReportedExternally` in constructor logic
   - In `PluginManager.js` apply telemetry handling (as it's done currently) only if `!serverless.isTelemetryReportedExternally` (it's to handle the case, where we have a fallback to local from global version which is older)
   - Move it right after we [either show help or run command in `scripts/serverless.js`](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/scripts/serverless.js#L613).
   - Optimize telemetry handling, and do not attempt to generate payload and call any telemetry related command if telemetry is disabled (currently we handle that setting in `storeLocally` to which already generated payload is send, while generation of payload involves some _async_ operations, it might be nice to avoid doing that as well).
   - Introduce a hack which ensures that eventual locally resolved `servelress` instance (which could come form an older version of a Framework) does not store or send _telemetry_ report on its own. For that add following:
     
      ```javascript
      require('../lib/utils/telemetry/areDisabled'); // Ensure value is resolved
      process.env.SLS_TRACKING_DISABLED = 1
      ```
      right before [`serverless.init()`](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/scripts/serverless.js#L373)
      Add also there a reasoning explanation and `TODO` comment indicating to remove that hack with next major (as with it, fallback to local version will happen right at begin of process handling)
1. [x] Make `serverless` argument optional in [lib/utils/telemetry/generatePayload.js](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/lib/utils/telemetry/generatePayload.js#L75):
   - Resolve `command` by calling `lib/cli/resolve-input` and not `serverless.processedInput` (note that interactive CLI command will now be resolved as`''`, and I think it's best if we leave it that way)
   - If no `serverless` passed, we can assume `serverless._isInvokedByGlobalInstallation` as `false`
   - If no `serverless` is passed resolved `isLocallyInstalled` in same manner as in [lib/cli/handle-error.js](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/lib/cli/handle-error.js#L106)
   - Assume no _service_ context if no `serverless` is passed
1. [x] Refactor [lib/cli/handle-error.js](https://github.com/serverless/serverless/blob/182d563da958670872962d437dd429e63a36989e/lib/cli/handle-error.js), so instead of taking `isLocallyInstalled` and `isInvokedByGlobalInstallation` options, it takes `serverless` instance as an option
1. [ ] Report whether command was successful, and report additional error information:
   - In `scripts/severless.js`:
      - Add to generated telemetry payload `outcome: "success"` property
      - Introduce top level local `hasTelemetryBeenReported` variable. Pass it with option under same name to [`handleError`](https://github.com/serverless/serverless/blob/154351f1a5925a745873895014ed31f03b2842b3/scripts/serverless.js#L23-L25) and set it to true, right prior report telemetry logic
   -  In [lib/cli/handle-error.js](https://github.com/serverless/serverless/blob/182d563da958670872962d437dd429e63a36989e/lib/cli/handle-error.js) by the end of the execution, if _telemetry_ is enabled, and `!options.hasTelemetryBeenReported` generate payload and store it and send it (unconditionally in that case).
      - Let's ignore the server response, and do not show any proposed notification.
      - Add  `outcome: "failure"`property to the payload
      - Add `failureReason` object property to the payload, with following data:
         - `kind` - If `isUserError` report as `user` otherwise as `programmer`
         - `code` - report `error.code` if it's found
         - `location` (only if `!isUserError || !error.code`) assign a truncated stack trace. How to resolve it (?) For a starting point take logic in this gist: https://gist.github.com/medikoo/a5c5223d69cf7cd80c0a5039cd4ee1ea (it's a method I've used to identify errors in error reporter I once was implementing for other project, and it seemed work well). One caveat is that, it was used for errors as happen in lambdas, so reporting full file paths from stack traces was harmless. Here we should rather truncate the location at some point. Probably simple approach would be to resolve common paths for all file paths in taken lines, and truncate it from each line. Note that it can be an error without any lines (e.g. Node.js fs tends to throw such, then we can safely report full stack track - I don't remember those errors reporting any user data in messages).
      - In case of _uncaught_ exception postpone the crash until telemetry request succeeds

**Comment by medikoo** [comment]
@pgrzesik let me know what you think about that

**Comment by pgrzesik** [comment]
Thanks a lot for the proposal @medikoo - I have some follow-up comments questions to the proposed approach:

> Move await storeTelemetryLocally call so it happens after command is processed (so to point in which we propagate eventual notification. Prior calling it, mark that we've send telemetry for given command by setting serverless._isTelemetryReported = true. Here add to the payload outcome: "success" property

What about the timing of actually sending the telemetry events? Should they still be sent in that flow, but only after `deploy` command?

> (not specific to this issue) Optimize telemetry handling, and do not attempt to generate payload and call any telemetry related command if telemetry is disabled (currently we handle that setting in storeLocally to which already generated payload is send, while generation of payload involves some async operations, it might be nice to avoid doing that as well).

Sounds like a great idea :+1: 

> Firstly with separate PR, ensure that all ServerlessError constructs, both in this repository and in @serverless/enterprise-plugin provide a meaningful error code as second argument, also verify that currently used codes are good.

Great call :+1: 

> In lib/cli/handle-error.js if we have serverless instance passed in options, there's no serverless._isTelemetryReported and telemetry is enabled, generate payload and store it and send it (unconditionally in that case).

How should the `send` work in this case? Should we still try to push all "cached" events with that report as well? 

Additionally - in this situation, we should not handle notifications at all, correct?

When it comes to reporting errors, what about inspecting the `providerError` which is ofter thrown by AWS? Maybe we could try to get the `code` of such error? It should not include any information but would give us insight of what actually failed on AWS (most likely on CloudFormation) side.

As for the `location` resolution - I need to understand a bit more how we can make sure that no sensitive information will end up in these before we can safely pass them in telemetry events. 


In addition to that, I believe we will have to adjust backend notification logic to only report them for successful deploy commands (do not process them for failed reports).

**Comment by medikoo** [comment]
> What about the timing of actually sending the telemetry events? Should they still be sent in that flow, but only after deploy command?

Yes, it's about moving all this logic to point after command is fully processed

> How should the send work in this case? Should we still try to push all "cached" events with that report as well?

I believe it should work exactly same as in non-error flow.

> Additionally - in this situation, we should not handle notifications at all, correct?

Correct, I forgot to mention that. Updated the spec

> When it comes to reporting errors, what about inspecting the providerError which is ofter thrown by AWS? Maybe we could try to get the code of such error?

I think best if implementation is not too complex. Maybe we can look at specific cases and discuss on that?

Ideally if `error.code` has all information we need (maybe we can propagate there `providerError.code` ?). Still if for some reason we want to keep both codes different, then we may consider sending the other code as well. Do you have example error case we can discuss on?

> As for the location resolution - I need to understand a bit more how we can make sure that no sensitive information will end up in these before we can safely pass them in telemetry events.

In what I proposed, we just share a few first  stack trace lines with file paths trimmed (so they're not absolute against system root). It'll share few first function names at which command crashed, but no arguments with which it was invoked. Also we won't share error message which may list some user input.

> In addition to that, I believe we will have to adjust backend notification logic to only report them for successful deploy commands (do not process them for failed reports).

We may, simply for optimization, but otherwise our in CLI logic should not attempt to show eventually returned notification

**Comment by medikoo** [comment]
@pgrzesik I've just realized that errors may be reported before the `Serverless` instance is created, e.g. some commands and variables validation is made earlier.

I've updated the spec so it provides a solution, so we can send telemetry also in such scenarios

**Comment by pgrzesik** [comment]
> Ideally if error.code has all information we need (maybe we can propagate there providerError.code ?). Still if for some reason we want to keep both codes different, then we may consider sending the other code as well. Do you have example error case we can discuss on?

One example was with `ECR` where initially we didn't consider that users should not be required to have ECR permissions if they're not using it: https://github.com/serverless/serverless/issues/8892 and later fix relies on specific `providerError.code` to avoid throwing in such case. But as you can see in example, it's anwyay handler as `ServerlessError`, so we can probably skip including these `providerError.code` codes separately. 

The proposal looks good, I don't have any further comments related to implementation. I will start with ensuring the `ServerlessError` instances have meaningful `code` associated with them.

**Comment by medikoo** [comment]
@pgrzesik I was reviewing now further interactive CLI improvements proposal, and realized that with this update, we actually have to move out telemetry handling of out `Serverless` class context entirely.

Thing is that interactive setup in the end will invoke multiple commands programmatically (e.g. `create` and then `deploy`), and we should not treat each as individual telemetry event in such case.

I've updated the spec above so telemetry reporting is fully moved to `scripts/serverless.js`. Also as it got more complicated I deciced to split it into few more PR's (5 in total in context of this repository), so we do not get bitten by too much complexity involved.

Please review again, and let me know what you think

**Comment by pgrzesik** [comment]
Thanks for revising it with `interactive` flow revamp in mind. I have one question related to second point. If we move the logic to `scripts` and we have a situation where global version is older (e.g. current `2.37.1`, without implementation in `scripts`) and local already has handling moved to scripts, the telemetry won't be reported at all in such setting - is my thinking correct here? I just wanted to confirm, I think it's not a major problem or anything, as this situation probably won't be very common, but it's something to keep in mind.

**Comment by pgrzesik** [comment]
In the meantime, I've reviewed `ServerlessError` instances in `enterprise-plugin` and looks like all of them have `code` already set.

**Comment by medikoo** [comment]
> If we move the logic to scripts and we have a situation where global version is older 

That's a very good point. Previously I've proposed a patch to handle the case, when it's the other way (global is new, but local is older), but that one was not envisioned. I've just updated the spec, where I proposed to pass `isTelemetryReportedExternally` option to constructor to handle this case.

**Comment by pgrzesik** [comment]
Sounds good to me :+1:
